bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/values/TableValueImpl.java

```diff
@@ -632,15 +635,39 @@ public V putData(V data) {
             MapValue dataMap = (MapValue) data;
             checkInherentTypeViolation(dataMap, tableType);
             K key = this.keyWrapper.wrapKey(dataMap);
+            Map.Entry<K, V> entry = new AbstractMap.SimpleEntry<>(key, data);
+            Long hash = TableUtils.hash(key, null);
+
+            if (entries.containsKey(hash)) {
+                return handleHashCollisionInPut(key, data, entry, hash);
+            }
 
             ArrayList<V> newData = new ArrayList<>();
             newData.add(data);
 
-            Map.Entry<K, V> entry = new AbstractMap.SimpleEntry<>(key, data);
-            Long hash = TableUtils.hash(key, null);
             return putData((K) key, data, newData, entry, hash);
         }
 
+        private V handleHashCollisionInPut(K key, V data, Map.Entry<K, V> entry, Long hash) {
+            updateIndexKeyMappings(hash, key, data);
+            List<Map.Entry<K, V>> entryList = entries.get(hash);
+            for (Map.Entry<K, V> extEntry: entryList) {
+                if (TypeChecker.isEqual(key, extEntry.getKey())) {
+                    entryList.remove(extEntry);
+                    entryList.add(entry);
+                    List<V> valueList = values.get(hash);
+                    valueList.remove(extEntry.getValue());
+                    valueList.add(data);
+                    return data;
+                }
+            }
+            List<Map.Entry<K, V>> extEntries = entries.get(hash);
```

Comment:
Is there a reason why we are getting entries with the same `hash` twice (`entryList` and `extEntries`), if the above loop modified `entryList` we shouldn't get here?